rfcs/2021-09-21-9292-throttle-transform.md

```diff
@@ -0,0 +1,153 @@
+# RFC 9292 - 2021-09-21 - `throttle` transform
+
+This RFC proposes the addition of a new transform that provides a user the ability to control the throughput of specific event streams.
+
+## Context
+
+- [#258](https://github.com/vectordotdev/vector/issues/258)
+
+## Scope
+
+### In scope
+
+- Dropping events (logs and metrics) to rate limit an event stream
+- Rate limit by both bytes (unserialized) and events
+- Exclude events with VRL conditions
+- Specify buckets based on the value of a key in the event
+
+### Out of scope
+
+- Apply backpressure as a form of rate limiting
```

Comment:
I think we should consider this in scope. Can you expand on why this is out of scope?

Reply:
I could be wrong, or missing a possible implementation, but I think this would be rather difficult to achieve with the way our topology is wired up currently where all events are sent along a single channel between components and back pressure is applied by the receiving side consuming more slowly. That is, you want to apply backpressure to separate "streams" of events, but they are all coming over the same channel.

Some half-baked ideas about how we could do this:

* Create a buffer per field value to feed events into when they are being throttled. This buffer could grow indefinitely though so we'd likely need to come up with some ways to limit growth that would eventually start shedding events anyway .
* At runtime, create a new channel per seen field value that is somehow fed all the way back to the source components so that back pressure could be applied to that channel separately. How would this work for sources where a given payload contains events that are throttled separately though?

I'd suggest we leave this for future work and just implement dropping events for now.

Reply:
Echoing what @jszwedko said, dropping events is much more straightforward, and much less invasive than attempting backpressure. I think DLQ would super useful here to allow events to be re-processed when throughput is lower.